llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```diff
@@ -3264,9 +3264,15 @@ bool SPIRVInstructionSelector::selectGlobalValue(
     PointerBaseType = GR.getOrCreateSPIRVType(
         GVType, MIRBuilder, SPIRV::AccessQualifier::ReadWrite, false);
   }
-  SPIRVType *ResType = GR.getOrCreateSPIRVPointerType(
-      PointerBaseType, I, TII,
-      addressSpaceToStorageClass(GV->getAddressSpace(), STI));
+
+  const unsigned AddrSpace = GV->getAddressSpace();
+  SPIRV::StorageClass::StorageClass StorageClass =
+      addressSpaceToStorageClass(AddrSpace, STI);
+  if (StorageClass == SPIRV::StorageClass::Function)
+    StorageClass = SPIRV::StorageClass::Private;
```

Comment:
This is the tricky part. In some ways it makes sense. A global value then that is in `Function` scope is meaningless. The choice are that we assert or change the storage class. I don't know how that will affect the code gen for functions.

If we do change the storage class to private, is there any documentation that will need to be updated?

Could this adhoc change here cause problems if another locations calls `addressSpaceToStorageClass` to get the storage class? If they don't do the same switch could we get inconsistent results. For example, a function that returns the pointer to a variable? At the call site, they will only see an addressspace of 0. How will it know if it actually returns value is a global value (Private) or something else (Function). I'm just thinking off the top of my head. This needs to be carefully thought out.

Reply:
Is there a way to adjust frontend to emit the proper address space for private variables (outside of function scope) before going to the backend? It would be the cleanest solution.

Reply:
This is indeed possible (initial PR added a `vulkan_private` address space).
But if the proposed RFC to set AS0=generic doesn't come true, I'd tend to believe AS0+global should end up in the private class, like AS0+local ends up in the Function class.

Reply:
I agree with concerns stated above by @s-perron , and as a part of instruction selection this substitution potentially makes the code more fragile and prone to error. The substitution itself however makes sense, so I think we can just move it to a later stage, when SPIR-V code is (almost) generated. Probably a more safe place for this substitution is SPIRVTargetLowering::finalizeLowering() where we keep all post-generation validation actions.

Additional benefit would be that all these kind of things are concentrated in one place, making post-generation validation a bit self-documented, and since this substitution logically belongs to the same sort of actions, its presence in SPIRVTargetLowering::finalizeLowering() among similar cases is more obvious and less unexpected for a developer.

Reply:
Actually, there is even better and simpler approach, similar to what @MrSidims asked about adjusting a frontend. We can do the substitution much much earlier, *before* IR Translation, to be sure that this is not intertwined with type inference and duplicates tracker.

Reply:
E.g., something along the lines of

```
bool SPIRVEmitIntrinsics::runOnModule(Module &M) {
  bool Changed = false;

  for (auto GV = M.global_begin(), GE = M.global_end(); GV != GE; ++GV) {
    if (isa<Function>(GV) || GV->getAddressSpace() != storageClassToAddressSpace(SPIRV::StorageClass::Function))
      continue;
    // add to worklist
  }
  // traverse worklist, re-create the global in another address space GV->replaceAllUsesWith(newGV);
```